fix(upgrade): silent bootstrap failures (#12062)

btrigueiro opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
#12062

**What is the new behavior?**
`bootstrap` now rejects the promise on errors and delegates the error handling to a common `onError` util (effectively logging the error and throwing it back)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

vicb committed:
refactor(core): simplify & cleanup reflection

vicb committed:
refactor(core): cleanup decorators

vicb committed:
fix(core): fix property decorators

fixes #12224

vicb committed:
fix(core): fix decorator defalut values

btrigueiro committed:
fix(upgrade): silent bootstrap failures (#12062)

vicb committed:
refactor(core): delete unused reflector code

matsko committed:
refactor(animations): ensure animation data-structures are created only when used

Closes #12250

vicb commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -418,7 +418,8 @@ export class UpgradeAdapter {
                     })
                     .then(resolve, reject);
-              });
+              })
```
could be `}, reject);` here

could you add a test ?

btrigueiro replied:
I'll change que `catch` to a `}, reject`. I gave it a quick try on the tests but as `upgrade`'s bootstrap is async and doesn't return promises I couldn't catch the thrown error. Maybe by attaching to some sort of zone's done event... But I'll need to dig a little on how to test it.

Ideally `upgradeAdapter.bootstrap` would be refactored so it returns promises like other bootstraps do. It would keep consistency and make testing a lot easier and cleaner. This is a little out of scope for this quick patch though. Let me know what you thing and we can work something out.

tbosch committed:
refactor(compiler): introduce directive wrappers to generate less code

- for now only wraps the `@Input` properties and calls
  to `ngOnInit`, `ngDoCheck` and `ngOnChanges` of directives.
- also groups eval sources by NgModule.

Part of #11683

michaelgeorgeattard committed:
refactor(common): remove redundant type

michaelgeorgeattard committed:
docs(ngIf): minor improvements

fank committed:
test(forms): added missing selfOnly tests (#12317)

gkalpak committed:
docs(CHANGELOG.md): remove incorrect link (#12259)

JohnAnthony committed:
docs(NgSwitch): fix mismatched tags in example (#12270)

gkalpak committed:
docs(CONTRIBUTING.md): fix typo (reproductions --> reproduction) (#12230)

fmalcher committed:
docs(router): attribute notation for string paths (#12205)

Either bind an expression or use the attribute notation.
The mixed way `[routerLink]="/path"` won't work.
Prefer the attribute notation for string-only paths

Commit:
refactor(http): use Http.request for all http shorthand methods (#12319)

robwormald committed:
fix(router): correctly export filter operator in es5 (#12286)

tbosch committed:
refactor(compiler): remove private exports

All of `@angular/compiler` is private, so we can export
everything we need directly.

tbosch committed:
refactor(compiler): don’t use the `OfflineCompiler` in extract_i18n

mprobst committed:
fix(http): make normalizeMethodName optimizer-compatible. (#12370)

`normalizeMethodName` reflectively accessed the RequestMethod enum. With a smart
optimizer, properties from the enum could be removed or renamed, and so user
code just passing in e.g. 'PATCH' might not work. This change fixes the code to
be more explicit and avoids the optimizer issue.

tbosch committed:
fix(benchmarks): fix method name in targetable spec

tbosch committed:
feat(benchmark): add large form benchmark

This benchmark tracks the generated file size for large forms
as well as the time to create and destroy many form fields.

matsko committed:
fix(animations): generate aot code for animation trigger output events (#12291)

Closes #11707
Closes #12291

alexeagle committed:
fix(compiler): don't redeclare a var in the same scope (#12386)

This is not allowed by Closure Compiler.

petebacondarwin committed:
chore(tslint.json): semicolon rule expects an array

Commit:
feat(forms): Validator.pattern accepts a RegExp (#12323)

cexbrayat committed:
feat(forms): add hasError and getError to AbstractControlDirective (#11985)

Allows cleaner expressions in template-driven forms.

Before:

    <label>Username</label><input name="username" ngModel required #username="ngModel">
    <div *ngIf="username.dirty && username.control.hasError('required')">Username is required</div>

After:

    <label>Username</label><input name="username" ngModel required #username="ngModel">
    <div *ngIf="username.dirty && username.hasError('required')">Username is required</div>

Fixes #7255

fank committed:
feat(forms): Added emitEvent to AbstractControl methods (#11949)

* feat(forms): Added emitEvent to AbstractControl methods

* style(forms): unified named parameter

Commit:
feat(forms): make 'parent' a public property of 'AbstractControl' (#11855)

pkozlowski-opensource committed:
feat(forms): add ng-pending CSS class during async validation (#11243)

Closes #10336

laskoviymishka committed:
feat(datePipe): support narrow forms for month and weekdays (#12297)

Closes #12294

AlmogShaul committed:
feat(forms) range values need to be numbers instead of strings (#11792)

alxhub committed:
chore(ci): make browserstack tests optional until they are fixed

vsavkin committed:
fix(router): fix lazy loading triggered by redirects from wildcard routes

Closes #12183

petebacondarwin committed:
feat(ngUpgrade): add support for AoT compiled upgrade applications

This commit introduces a new API to the ngUpgrade module, which is compatible
with AoT compilation. Primarily, it removes the dependency on reflection
over the Angular 2 metadata by introducing an API where this information
is explicitly defined, in the source code, in a way that is not lost through
AoT compilation.

This commit is a collaboration between @mhevery (who provided the original
design of the API); @gkalpak & @petebacondarwin (who implemented the
API and migrated the specs from the original ngUpgrade tests) and @alexeagle
(who provided input and review).

This commit is an starting point, there is still work to be done:

* add more documentation
* validate the API via internal projects
* align the ngUpgrade compilation of A1 directives closer to the real A1
  compiler
* add more unit tests
* consider support for async `templateUrl` A1 upgraded components

Closes #12239

vicb committed:
refactor: remove some facades (#12335)

vsavkin committed:
cleanup(router): add a test verifying than NavigationEnd is not emitted after NavigationCancel

alxhub committed:
chore(ci): re-enable browserstack tests in ci

vsavkin committed:
feat(router): add support for ng1/ng2 migration (#12160)

vsavkin committed:
fix(router): module loader should start compiling modules when stubbedModules are set (#11742)

mucer committed:
fix(router): do not update primary route if only secondary outlet is given (#11797)

IgorMinar committed:
chore(release): cut the 2.2.0-beta.0 release and add changelog

tbosch committed:
fix(compiler): don't access view local variables nor pipes in host expressions (#12396)

Fixes #12004
Closes #12071

IgorMinar committed:
docs(changelog): add 2.1.1 changelog

andreialecu commented:
I think you might need to rebase this on the current master instead of all this commits, 

Maintainers, please try to get this merged ASAP, it makes working with `@angular/upgrade` impossible otherwise.

https://github.com/angular/angular/issues/12062#issuecomment-255582935

andreialecu commented:
Actually, this doesn't seem to fix all cases.

Having an invalid element does indeed make this show the error, but try adding a `&dash;` somewhere in the template and you will still see it fail silently.

I wonder if there are additional cases where that may happen.

andrei-ivanovici commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -418,7 +418,7 @@ export class UpgradeAdapter {
                     })
                     .then(resolve, reject);
-              });
+              }, reject);
```
This simplification appears to  break the  error handling again. Using the original solution  with the catch fixes this. Can anyone else confirm
```
 }, reject);
```
whould be 
```
 }).catch(reject)
```

andreialecu replied:
I can confirm, see https://github.com/angular/angular/issues/12062#issuecomment-255585965
